---
title: "Practical 2: The epigenetic clock: Variable ranking and multiple testing"
author: "Verena Zuber, Jelena Besevic,  Alpha Forna, and Saredo Said"
date: "31/1/2019"
output: pdf_document
highlight: tango
---


## Part 1: The epigenetic clock: Epigenetic marks associated with ageing


Our epigenome is highly impacted by environmental changes. One particular interesting aspect is ageing and how epigentic marks such as methylation are affected by ageing. Scientists have shown that there exist specific methylation sites that correlate with age, an observation that has been made in humans, chimpanzees, mice, or rats. Based on our knowledge which methylation sites correlate with healthy ageing we can use these epigenetic marks as biomarkers to predict the  "actual biological age" of an individual. For example, if an individual has been exposed to pollutants or suffered from stress, its "actual biological age" of its body might be much older than its true age.

If you want to read more on the topic, there is a Nature Feature on the scientist  Steve Horvath who first proposed to use methylation to measure the biological age

https://www.nature.com/news/biomarkers-and-ageing-the-clock-watcher-1.15014

In this practical we consider data on $n=409$ healthy mice and methylation of $p=3,663$ conserved methylation sites. Load the dataset, that contains the methylation matrix as predictor matrix and the age of the mice (in months) stored in the vector y. Familiarise yourself with the dataset using the following commands

```{r message=FALSE, warning=FALSE, fig.height=2.8}
load("data_epigenetic_clock_control")
#alternatively try load("data_epigenetic_clock_control.dms")
y = control_mice$y_control
hist(y,breaks=50, main="")
x = control_mice$x_control
dim(x)
```
The first part is concerned with performing a ranking of methylation sites that have the strongest association with ageing. 


Question 1.1

Compute a linear regression of the first methylation site against the age of the mice. Note in order to access the first column of a matrix, use square bracket like this [,1] and for the $j$th variable use [,j]. Figure out which element in the $\$$coefficients matrix contains the $p$-value of the regression coefficient. Use again the squared brackets to index only the $p-$value. 

*Reply: We fit the regression using the lm() function and look at the value  $\$$coefficients.*
```{r message=FALSE, warning=FALSE}
summary(lm(y~x[,1]))$coefficients
```
*The element in the 2nd row and the 4th columns of the $\$$ coefficient value contains the $p-$value.*

```{r message=FALSE, warning=FALSE}
summary(lm(y~x[,1]))$coefficients[2,4]
```

Question 1.2

In order to compute the massively univariate linear regression estimate, we need to automate this computation for all $p=3,663$ methylation sites. First initiate a vector where to save your $p-$values.
```{r message=FALSE, warning=FALSE}
pvec = rep(NaN, 3663)
```

Write a 'for loop' to iterate through all variables. In case you are not familiar with the 'for loop', use this practical here for help https://www.r-bloggers.com/how-to-write-the-first-for-loop-in-r/ . In each iteration $j$ save the $p-$value of the respective regression into the pvec vector at position pvec[j].


*Reply: This is how such a for loop could look like. *

```{r message=FALSE, warning=FALSE}
for(j in 1:ncol(x)){
	pvec[j]=summary(lm(y~x[,j]))$coefficients[2,4]
}
```

Question 1.3

Rank the methylation sites according to their $p-$values and show the top 10 methylation sites that are associated with ageing.

*Reply: First we combine the methylation names with the $p-$values and than we sort the output using the order function. *

```{r message=FALSE, warning=FALSE}
output=cbind(colnames(x),pvec)
output_sorted=output[order(pvec),]
head(output_sorted,n=10)
```


## Part 2: FDR control of the variable ranking

The next step is to correct the $p$-values for multiple testing. Since we performed $p=3,663$ tests we need to make sure we do control the rate of false positive discoveries. There are different approaches to correct for multiple testing. Here, we perform the key methods and discuss the results in the end. 

Question 2.1
Plot a histogram and discuss which distribution this vector of $p-$values has.

*Reply: We see here a mixture distribution of a uniform distribution which represents the Null variables and a steep peak at around zero which contains the signal variables. *

```{r message=FALSE, warning=FALSE, fig.height=2.8}
hist(pvec, breaks = 50, main="")
```

Question 2.2
Perform the Bonferroni correction on your $p-$value vector. Use $\alpha=0.05$ as your significance threshold. 
How many methylation sites would be considered as Non-Null after Bonferroni correction?

*Reply: Using Bonferroni correction we detect 7 methylation sites to be significant. *

```{r message=FALSE, warning=FALSE}
bonferroni = p.adjust(pvec, method="bonferroni")
table(bonferroni<0.05)
output[which(bonferroni<0.05),]
```


Question 2.3
Perform the Benjamini-Hochberg Fdr correction on your $p-$value vector. Use $\alpha=0.05$ as your significance threshold. 
 How many methylation sites would be considered as Non-Null after Benjamini-Hochberg Fdr correction?

*Reply: Using Benjamini Hochberg correction we detect 35 methylation sites to be significant. *

```{r message=FALSE, warning=FALSE}
bh = p.adjust(pvec, method="BH")
table(bh<0.05)
#output[which(bh<0.05),]
```

Question 2.4
Perform the $q$-value Fdr correction on your $p-$value vector. The qvalue function is provided in the qvalue package on Bioconductor. Make sure you have the latest version of R (>3.5) installed. Follow the instructions on the homepage for installation https://www.bioconductor.org/packages/release/bioc/html/qvalue.html and call the package using  
```{r message=FALSE, warning=FALSE}
library(qvalue)
```
Use $\alpha=0.05$ as your significance threshold.  How many methylation sites would be considered as Non-Null after $q-$value Fdr correction?

*Reply: Using the $q-$value correction we detect 39 methylation sites to be significant. 
```{r message=FALSE, warning=FALSE}
qobj = qvalue(p=pvec)
qvalues = qobj$qvalues
table(qvalues<0.05)
#output[which(qvalues<0.05),]
```
*Further we see that the estimated proportion of Null variables (pi0) is 0.753487, which is equivalent to (1-0.753487) = 0.246513 Non-Null variables. Given that the dataset contains 3663 variables this would be around 900 potential signal variables. This indicates that there is a strong signal and many methylation sites are potentially associated with ageing, but given our sample size of n=409 we can only detect 39 methylation sites. *
```{r message=FALSE, warning=FALSE}
qobj$pi0
```



Question 2.5
Perform the local fdr correction on your $p-$value vector using the fdrtool R package. Use the option statistic='pvalue' to model the $p-$value vector and extract the local fdr estimate in the $\$lfdr$ value. How many methylation sites would be considered as Non-Null after local fdr correction at a level of $\alpha=0.2$?

*Reply: Using the local fdr correction we declare 164 methylation sites as discoveries.*
```{r message=FALSE, warning=FALSE, fig.height=7}
library(fdrtool)
lfdr_out = fdrtool(x=pvec, statistic="pvalue", verbose=FALSE)
table(lfdr_out$lfdr<0.2)
#output[which(lfdr_out$lfdr<0.05),]
```

Question 2.6
Look at the top graph of the mixture models that the fdrtool package has as ouput. Where is the Null-distribution and where is the Non-Null distribution? Would you consider this to be a good model fit? 

*Reply: In red we can see the Null distribution which is a uniform distribution and in blue we see the Non-Null distribution. The underlying histogram shows the actual observed data. We indeed see a good fit of the observed data with our fitted mixture distribution. *




Question 2.7
The fdrtool estimates the proportion of Null variables. Access this information in the $\$$param value. What is the proportion of Null variables?

*Reply: The estimated proportion of Null variables (pi0) is 0.7677763, which is equivalent to (1-0.753487) = 0.2322237 Non-Null variables. *

```{r message=FALSE, warning=FALSE}
lfdr_out$param
```


Question 2.8
Compare the 4 ways of how to perform multiple testing correction and discuss which one is the most conservative.

*Reply: The most conservative approach is the Bonferroni correction. Regarding the FDR approaches, the Benjamini-Hochberg approach is more conservative than the $q-$value, since it assumes that a priori there is no signal $(pi0=1)$. In contrast, the $q-$value estimates $pi0$ from the data, allowing $pi0<1$; in this particular analysis the propostion of Null variables is 0.753487. These tail-area Fdr estimates are essentially adjusted $p-$values, while the local fdr is the posterior that a variable is Null given its observed $p-$value.  *






## Part 3 (optional): Random p-values

In order to evaluate the Null-distribution of $p$-values of a linear regression model we perform a simulation study


Question 3.1
First simulate both x and y from the normal-distribution with mean 0 and variance 1, for each variable draw a 1000 random draws using the rnorm() function in R  using
```{r message=FALSE, warning=FALSE}
set.seed(1234)
x=rnorm(1000)
y=rnorm(1000)
```
Compute the linear regression of $x$ on $y$ and look at the $p$-value of the regression coefficient.

*Reply: We fit the regression using the lm() function and look at the value  $\$$coefficients. *
```{r message=FALSE, warning=FALSE}
summary(lm(y~x))$coefficients
```



Question 3.2
Repeat this operation 10,000 times using a for loop. Save your $p-$values of each iteration in an object pvec defined as
```{r message=FALSE, warning=FALSE}
pvec = rep(NaN, 10000)
```
*Reply: We do this with the following for loop*

```{r message=FALSE, warning=FALSE}
for(i in 1:10000){

	x=rnorm(1000)
	y=rnorm(1000)
	pvec[i]=summary(lm(y~x))$coefficients[2,4]

}

```


Question 3.3
Plot a histogram and discuss which distribution this vector of $p-$values has.
```{r message=FALSE, warning=FALSE, fig.height=3}
hist(pvec, breaks=50, main="")
```

*Reply: The Null-distribution of $p-$values follows a uniform distribution between 0 and 1.*






## Part 4 (Optional): FDR correction on GWAS summary data on Schizophrenia

Go to the webpage of the Psychiatric Genetics Consortium (PGC) https://www.med.unc.edu/pgc/results-and-downloads
and download in the section SCZ2 the full SNP results by clicking on the link 'Download full SNP results'.
After downloading and un-zipping the file we can load it into R using and check the dimension of the dataset
```{r message=FALSE, warning=FALSE}
data=read.csv("ckqny.scz2snpres", sep="\t")
dim(data)
```
The file includes genome-wide data on over 9m SNPs. Looking at the first few lines of the data we find the column providing the $p$-values for the SNPS and save it as a vector pvec.
```{r message=FALSE, warning=FALSE}
head(data)
pvec = data$p
N=length(pvec)
N
```


Question 4.1
Plot a histogram of the $p$-value vector pvec. Discuss if this distribution can be modelled using a mixture distribution.

*Reply: This is a perfect mixture distribution that consists of the following two distributions: 
- $F_0$ a uniform distribution of the Null or irrelavant SNPs and
- $F_A$ the sharp peak at around 0 that captures the Non-Null SNPs or the signal.*

```{r message=FALSE, warning=FALSE, fig.height=2.8}
hist(pvec, breaks=100, main="")
```



Question 4.2
Perform Bonferroni correction of the p-value vector pvec using the function p.adjust. How many SNPs are significant at a Bonferroni adjusted level  of 0.05?


*Reply: Following the Bonferroni correction there are 9,323 SNPs discovered at a level of 0.05.*

```{r message=FALSE, warning=FALSE}
p_bonferroni=p.adjust(p=pvec, method="bonferroni")
table(p_bonferroni<0.05)
```



Question 4.3
Perform the Benjamini-Hochberg FDR correction of the p-value vector pvec using the function p.adjust. How many SNPs are significant at a Benjamini Hochberg FDR adjusted level  of 0.05?

*Reply: Following the Benjamini-Hochberg correction there are 12,4955 SNPs discovered at a level of 0.05.*

```{r message=FALSE, warning=FALSE}
p_bh=p.adjust(p=pvec, method="BH")
table(p_bh<0.05)
```



Question 4.3
Perform the q-value FDR correction of the p-value vector pvec using the function qvalue in the qvalue package on bioconductor. Follow the instructions on the homepage for installation https://www.bioconductor.org/packages/release/bioc/html/qvalue.html and call the package using  
```{r message=FALSE, warning=FALSE}
library(qvalue)
```
How many SNPs are significant at a $q-$value level of 0.05? What is the proportion of Null SNPs?

*Reply: Following the $q-$value correction there are 141,181 SNPs discovered at a level of 0.05.*

```{r message=FALSE, warning=FALSE}
qobj = qvalue(p=pvec)
qvalues = qobj$qvalues
table(qvalues<0.05)
```


*Looking at the pi0 object we see that the actual proportion of Null-SNPs is 0.8394974, which suggests that around 16$\%$ percent of the SNPs are associated with schizophrenia. This suggests that schizophrenia is a highly polygenic disease.*

```{r message=FALSE, warning=FALSE}
pi0 = qobj$pi0
pi0
```


Question 4.4
Perform the local fdr correction on your $p-$value vector using the fdrtool R package. Use the option statistic=`pvalue' and extract the local fdr estimate in the $\$lfdr$ value. How many SNPs would be considered as Non-Null after local fdr correction at a level of $\alpha=0.2$?


*Reply: At a local fdr level of $0.2$ there are 212,322 SNPs declared as discoveries.*

```{r message=FALSE, warning=FALSE, fig.height=7}
library(fdrtool)
lfdr_out = fdrtool(x=pvec, statistic="pvalue", verbose=FALSE)
table(lfdr_out$lfdr<0.2)
```



Question 4.5
Look at the graph of the mixture models that the fdrtool package has as ouput. Would you consider this to be a good model fit? Further, what is the proportion of Null variables?


*Reply: The observed data in the histogram fits nicely the fitted mixture model in black consisting of the red Null and blue Non-Null component.*
```{r message=FALSE, warning=FALSE}
lfdr_out$param
```

*If you want to read more on the implications of FDR analysis in genome-wide association studies, take a look at the following recent article 'Association analyses based on false discovery rate implicate new loci for coronary artery disease' in Nature Genetics
https://www.nature.com/articles/ng.3913 
If you are accessing this article from outside of the College:
- Click the link
- Select Shibboleth Under 'Additional access options'
- Search for and select Imperial College London
- Login with your College username and password
*



<!--
require(knitr)
require(markdown)
require(rmarkdown)
render("practical2_solutions.Rmd")
-->

